src/main/java/mops/businesslogic/FileServiceImpl.java

```
@@ -0,0 +1,213 @@
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import mops.businesslogic.exception.DeleteAccessPermissionException;
+import mops.businesslogic.exception.FileNotFoundException;
+import mops.businesslogic.exception.ReadAccessPermissionException;
+import mops.businesslogic.exception.WriteAccessPermissionException;
+import mops.exception.MopsException;
+import mops.persistence.FileRepository;
+import mops.persistence.directory.Directory;
+import mops.persistence.file.FileInfo;
+import org.springframework.core.io.ByteArrayResource;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.transaction.interceptor.TransactionAspectSupport;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Set;
+
+@Service
+@AllArgsConstructor
+@Slf4j
+public class FileServiceImpl implements FileService {
+
+    /**
+     * Service for permission checks.
+     */
+    private final DirectoryService directoryService;
+    /**
+     * Service for saving and retrieving file meta data.
+     */
+    private final FileInfoService fileInfoService;
+    /**
+     * File content repository.
+     */
+    private final FileRepository fileRepository;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @SuppressWarnings("PMD.LawOfDemeter")
+    @Transactional(rollbackFor = MopsException.class)
+    public void saveFile(Account account, long dirId, MultipartFile multipartFile,
+                         Set<String> tags) throws MopsException {
+        UserPermission userPermission = directoryService.getPermissionsOfUser(account, dirId);
+
+        if (!userPermission.isWrite()) {
+            log.error("User {} tried to save a file without write permission.",
+                    account.getName(),
+                    new WriteAccessPermissionException("No write permission")
```

you need to throw this exception instead of giving it to the logger.